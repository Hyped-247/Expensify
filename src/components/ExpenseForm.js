import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';


// <!-- You can only autoFocus on input at a time. The first one. -->
export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ?  (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        };
    }

    OnDescriptionChange = (e) => {
       const description = e.target.value;
       this.setState(() => ({description}));
    };

    OnChangeTextArea = (e) => {
       const note = e.target.value;
       this.setState(() => ({note}));
    };

    OnChangeAmount = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}));
        }
    };

    OnDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({createdAt}))
        }
    };

    OnFocusChange = (focused) => {
        this.setState(() => ({calenderFocused: focused}))
    };

    OnSubmit = (e) => {
        e.preventDefault();
        const amount = this.state.amount;
        const description = this.state.description;
        if (!amount || !description) {
            this.setState(() => ({error: 'Please provide description and amount.'}))
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    // Without defining onChange the inputs will be read only.
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.OnSubmit}>
                    <input type="text"
                           placeholder="description"
                           autoFocus
                           className={"text-input"}
                           value={this.state.description}
                           onChange={this.OnDescriptionChange}
                    />
                    <input type="number"
                           placeholder="Amount"
                           className={"text-input"}
                           value={this.state.amount}
                           onChange={this.OnChangeAmount}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.OnDateChange}
                        focused={this.state.calenderFocused}
                        onFocusChange={this.OnFocusChange}
                        numberOfMonths={1} // this will determent the number of calenders we can see.
                        isOutsideRange={() => false} // this will make us pick days in the past.
                    />

                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        className={"textarea"}
                        onChange={this.OnChangeTextArea}>
                    </textarea>

                    <button>
                        {this.props.expense ? <span>Update Expense</span> : <span>Add Expense</span>}
                    </button>
                </form>
            </div>
        )
    }
}
