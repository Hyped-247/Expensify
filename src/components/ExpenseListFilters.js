import React from 'react';
import {connect} from "react-redux";
import {setTextFilter, sortByDate,
        sortByAmount, setEndDate,
        setStartDate} from "../actions/filters";
import {DateRangePicker} from "react-dates";


export class ExpenseListFilters extends React.Component {

    state = {calendarFocused: null};

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({calendarFocused}))
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        const option = e.target.value;
        if (option === 'date') {
            this.props.sortByDate();
        } else {
            this.props.sortByAmount();
        }
    };

    render() {
        return (
            // update the store when you input change the data on the text field.
            <div className={"content-container"}>
                <div className={"input-group"}>
                    <div className={"input-group__item"}>
                         <input className={"text-input"} type="text" value={this.props.filters.text}
                                onChange={this.onTextChange} />
                    </div>
                    <div className={"input-group__item"}>
                         <select className={"select"} value={this.props.filters.sortBy}
                                 onChange={this.onSortChange}
                                 placeholder={"Search expenses"}>

                             <option value="date">Date</option>
                             <option value="amount">Amount</option>
                         </select>
                    </div>
                    <div className={"input-group__item"}>
                      <DateRangePicker startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false} />
                    </div>
                </div>
            </div>
            );
    }
}

const mapStateToProps = (state) => {
    return {filters: state.filters};
};

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);

