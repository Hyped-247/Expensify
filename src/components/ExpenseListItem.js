import React from 'react';
import {Link} from "react-router-dom";


const ExpenseListItem = (props) => (
    <div>
        <h2>This is my ExpenseList component.</h2>
        <div>
            <Link to={`/edit/${props.id}`}>
                <h2>{props.description}</h2>
            </Link>
            <h2>{props.amount}</h2>
            <h2>{props.createdAt}</h2>
        </div>
    </div>
);


export default ExpenseListItem;