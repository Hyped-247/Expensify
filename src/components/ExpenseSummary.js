import React from 'react';
import {connect} from 'react-redux';
import selectExpenses from "../selectors/expenses";
import {getExpensesTotal} from "../selectors/expenses-total";


export const ExpenseSummary = ({expenseCount, expensesTotal}) => (
    <div>
        <h2>Summary</h2>
        <h1>Total count of expenses: {expenseCount} And amount is: {expensesTotal}</h1>
    </div>
);


const mapStateToProps = (state) => {
    const filteredExpenses = selectExpenses(state.expenses, state.filters);
    return {
        expenseCount: filteredExpenses.length,
        expensesTotal: getExpensesTotal(filteredExpenses)
    };
};


export default connect(mapStateToProps)(ExpenseSummary);
