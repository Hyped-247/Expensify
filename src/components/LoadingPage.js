import React from "react";

const LoadingPage = () => (
    <div className={"loader"}>
        <img alt={""} className={"loader__img"} src={'public/images/loader.gif'} />
    </div>
);

export default LoadingPage;




