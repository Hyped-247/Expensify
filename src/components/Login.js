import React from 'react';
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";


export const Login = (props) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h2 className={"box-layout__title"}>Expensify Web Application</h2>
            <button className={"button"} onClick={props.startLogin}>Login</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(Login)
