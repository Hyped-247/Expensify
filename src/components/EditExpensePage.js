import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {startEditExpenses, startRemoveExpenses} from "../actions/expenses";


export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.startEditExpenses(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    onClick = () => {
        this.props.startRemoveExpenses({id: this.props.expense.id});
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
                <button onClick={this.onClick}>Delete</button>
            </div>
        );
    }
}



const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id
)});


const mapDispatchToProps = (dispatch, ) => {
    return {
        startEditExpenses: (id, expense) => dispatch(startEditExpenses(id, expense)),
        startRemoveExpenses: (data) => dispatch(startRemoveExpenses(data))
    };
};



export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
