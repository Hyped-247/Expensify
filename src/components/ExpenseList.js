import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                <h2>{props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />;
                })}</h2>
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        // selectExpenses is going to process the data before it gets used in the comp.
        expenses: selectExpenses(state.expenses, state.filters)
    }
};


export default connect(mapStateToProps)(ExpenseList);
