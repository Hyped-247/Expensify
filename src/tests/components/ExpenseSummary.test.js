import React from 'react';
import {shallow} from "enzyme";
import {ExpenseSummary} from "../../components/ExpenseSummary";


test('render Expense Summary with one expense', () => {
    const wrapper = shallow(<ExpenseSummary expensesTotal={1} expenseCount={333}/>);
    expect(wrapper).toMatchSnapshot();
});


test('render Expense Summary with one expense', () => {
    const wrapper = shallow(<ExpenseSummary expensesTotal={3} expenseCount={322222222233}/>);
    expect(wrapper).toMatchSnapshot();
});
