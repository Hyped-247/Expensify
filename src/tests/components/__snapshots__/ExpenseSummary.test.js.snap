// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Expense Summary with one expense 1`] = `
<div>
  <h2>
    Summary
  </h2>
  <h1>
    Total count of expenses: 
    333
     And amount is: 
    1
  </h1>
</div>
`;

exports[`render Expense Summary with one expense 2`] = `
<div>
  <h2>
    Summary
  </h2>
  <h1>
    Total count of expenses: 
    322222222233
     And amount is: 
    3
  </h1>
</div>
`;
